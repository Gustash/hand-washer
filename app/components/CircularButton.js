import React from 'react'
import { TouchableOpacity, View, StyleSheet, Text } from 'react-native'
import { colors } from 'app/config'

export const DEFAULT_CIRCULAR_BUTTON_SIZE = 204

const CircularButton = ({
  size = DEFAULT_CIRCULAR_BUTTON_SIZE,
  onPress,
  text,
  icon,
  style,
}) => (
  <TouchableOpacity
    style={[
      styles.container,
      { width: size, height: size, borderRadius: size / 2 },
      style,
    ]}
    onPress={onPress}
    disabled={!onPress}
  >
    <View style={styles.icon}>{icon}</View>
    <Text style={styles.text}>{text}</Text>
  </TouchableOpacity>
)

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.lightBlue,
    shadowColor: colors.black,
    shadowOpacity: 0.2,
    shadowOffset: {
      height: 2,
    },
    shadowRadius: 16,
    padding: 30,

    elevation: 16,
  },
  icon: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    color: colors.white,
    textAlign: 'center',
    fontSize: 21,
  },
})

export default CircularButton
