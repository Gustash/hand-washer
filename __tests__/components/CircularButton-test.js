import React from 'react'
import { Text, TouchableOpacity, StatusBar } from 'react-native'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import CircularButton from '../../app/components/CircularButton'

const mockProps = props => ({
  onPress: jest.fn(),
  text: '',
  icon: '',
  style: null,
  ...props,
})

describe('<CircularButton />', () => {
  let props = mockProps()
  let wrapper = shallow(<CircularButton {...props} />)

  it('matches CircularButton snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  describe('CircularButton applies default props', () => {
    it('defaults size', () => {
      expect(wrapper.find(TouchableOpacity).prop('style')).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            width: 204,
            height: 204,
            borderRadius: 102,
          }),
        ])
      )
    })
  })

  describe('CircularButton applies props', () => {
    beforeEach(() => {
      props = mockProps({
        text: 'Circular Button',
        style: {
          marginTop: 1000,
        },
        size: 64,
        icon: <StatusBar backgroundColor="#3a3a3a" />,
      })
      wrapper = shallow(<CircularButton {...props} />)
    })

    it('renders text', () => {
      expect(
        wrapper
          .find(Text)
          .first()
          .prop('children')
      ).toBe('Circular Button')
    })

    it('handles onPress', () => {
      wrapper.simulate('press')

      expect(wrapper.prop('onPress')).toHaveBeenCalledTimes(1)
    })

    it('disables Touchable without onPress', () => {
      expect(wrapper.find(TouchableOpacity).prop('disabled')).toBe(false)

      const propsWithoutOnPress = {
        ...wrapper.props(),
        onPress: null,
      }
      wrapper.setProps(propsWithoutOnPress)

      expect(wrapper.find(TouchableOpacity).prop('disabled')).toBe(true)
    })

    it('applies style', () => {
      expect(wrapper.find(TouchableOpacity).prop('style')).toContainEqual({
        marginTop: 1000,
      })
    })

    it('applies size', () => {
      expect(wrapper.find(TouchableOpacity).prop('style')).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            width: 64,
            height: 64,
            borderRadius: 32,
          }),
        ])
      )
    })

    it('renders icon', () => {
      expect(
        wrapper.containsMatchingElement(<StatusBar backgroundColor="#3a3a3a" />)
      ).toBe(true)
    })
  })
})
