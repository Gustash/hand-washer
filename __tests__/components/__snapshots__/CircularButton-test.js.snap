// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircularButton /> matches CircularButton snapshot 1`] = `
<ForwardRef
  disabled={false}
  onPress={[MockFunction]}
  style={
    Array [
      Object {
        "backgroundColor": "#00A1D2",
        "elevation": 16,
        "padding": 30,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 16,
      },
      Object {
        "borderRadius": 102,
        "height": 204,
        "width": 204,
      },
      null,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  />
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 21,
        "textAlign": "center",
      }
    }
  />
</ForwardRef>
`;
